### ISSUE TITLE
CheckReturnValue annotation applied to package is not enforced

### ISSUE THREAD

**Issue by RichardBradley** [spotbugs/spotbugs#429]
The `RV_RETURN_VALUE_IGNORED_BAD_PRACTICE` rule detects whether methods that are marked `@CheckReturnValue` have their return value ignored, and warns if so.

This should catch a class of bugs in fluent assertion libraries like "[AssertJ](http://joel-costigliola.github.io/assertj/)" and "[Google Truth](https://github.com/google/truth)", i.e.

```
assertThat(a == b); // BUG: this is a no-op

assertThat(a).isEqualto(b); // correct assertion
```

See the discussion in the AssertJ library about this issue: http://joel-costigliola.github.io/assertj/assertj-core.html#faq-incorrect-api-usage and https://github.com/joel-costigliola/assertj-core/pull/695

The AssertJ library has applied `@CheckReturnValue` to each of their methods, but the "Google Truth" library has applied `@CheckReturnValue` to their entire package: https://github.com/google/truth/blob/9cfceb957c0e42b76ca598c07900414081a81144/core/src/main/java/com/google/common/truth/package-info.java#L28

The former works with Findbugs/Spotbugs, but the latter is not detected. I think it should be, as the `javax.annotation.CheckReturnValue` annotation is allowed at the package level, so the designers clearly intended this kind of usage.

## Repro case:

Code like the following should warn with `RV_RETURN_VALUE_IGNORED_BAD_PRACTICE`, but do not:

```
import org.junit.*;
import static com.google.common.truth.Truth.assertThat;

public class MyTest {
  @Test
  public void testExample() {
    assertThat(System.currentTimeMillis() > 0); // BUG: this is a no-op
  }
}
```

**Comment by KengoTODA** [comment]
Reproduced this problem by 5ac739bbce0a1e76504b921d23499565557d52d6

**Comment by KengoTODA** [comment]
I found the reason:

`BuildCheckReturnAnnotationDatabase` (subclass of `AnnotationVisitor`) does not visit classes in AUX_CLASSPATH. `CheckReturnAnnotationDatabase` visits classes in AUX_CLASSPATH but it needs the Map named `defaultAnnotation` built by `BuildCheckReturnAnnotationDatabase`.

**Comment by cchesser** [comment]
I don't see an issue with the behavior described on this correction; however, I wanted to share an observation of what we are now seeing as a failure by using 3.1.0-RC7 (as they may be helpful for others). We have several references to guava's `com.google.common.collect` package, which [has `@CheckReturnValue` at the package level](https://github.com/google/guava/blob/v23.2/guava/src/com/google/common/collect/package-info.java#L221). Several APIs within these collect package (ex. [ImmutableSet](https://github.com/google/guava/blob/v23.2/guava/src/com/google/common/collect/ImmutableSet.java#L509)) use the error-prone annotation `@CanIgnoreReturnValue`. Therefore, the author's chose the share the fact that this can be ignored, but used an annotation which is specific to what is doing this static analysis: error-prone (most of the notes [can be found here](https://github.com/google/error-prone/issues/279)).

As a result, we are building out exclusion file for cases where matches can be ignored, for the case where they are documented as `@CanIgnoreReturnValue` with the error-prone annotation.

**Comment by KengoTODA** [comment]
Current spotbugs-annotation does not support this requirement. However [JSR305](https://github.com/amaembo/jsr-305/blob/master/ri/src/main/java/javax/annotation/CheckReturnValue.java) supports this, just annotate target methods by `@CheckReturnValue(when = When.NEVER)`.

#421 should cover this requirement in future.
